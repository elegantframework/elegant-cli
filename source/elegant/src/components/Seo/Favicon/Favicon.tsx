interface Props {
    /**
     * The directory path of the favicons
     */
    directory?: string;
    /**
     * Display an icon for apple devices?
     */
    appleTouchIcon?: boolean;
    /**
     * Display an icon for android devices?
     */
    androidIcon?: boolean;
    /**
     * Display favicons?
     */
    favicon?: boolean;
};

/**
 * Generate a complete list of favicon meta links
 * @returns A completes list of favicons for different device types
 */
const Favicon = ({
    directory = "/favicons",
    appleTouchIcon = true,
    androidIcon = true,
    favicon = true
}: Props) => {

    let icons = [];

    if(appleTouchIcon){
        icons.push(
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-57x57.png`,
                sizes: '57x57'
            },
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-60x60.png`,
                sizes: '60x60'
            },
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-72x72.png`,
                sizes: '72x72'
            },
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-76x76.png`,
                sizes: '76x76'
            },
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-114x114.png`,
                sizes: '114x114'
            },
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-120x120.png`,
                sizes: '120x120'
            },
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-144x144.png`,
                sizes: '144x144'
            },
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-152x152.png`,
                sizes: '152x152'
            },
            {
                rel: 'apple-touch-icon',
                href: `${directory}/apple-icon-180x180.png`,
                sizes: '180x180'
            } 
        );
    }

    if(androidIcon){
        icons.push(
            {
                rel: 'icon',
                href: `${directory}/android-icon-192x192.png`,
                sizes: '192x192'
            },
            {
                rel: 'manifest',
                href: `${directory}/manifest.json`
            }
        );
    }

    if(favicon){
        icons.push(
            {
                rel: 'icon',
                href: `${directory}/favicon-16x16.png`,
                sizes: '16x16'
            },
            {
                rel: 'icon',
                href: `${directory}/favicon-32x32.png`,
                sizes: '32x32'
            },
            {
                rel: 'icon',
                href: `${directory}/favicon-96x96.png`,
                sizes: '96x96'
            },
        );
    }

    return icons;
};

export default Favicon;