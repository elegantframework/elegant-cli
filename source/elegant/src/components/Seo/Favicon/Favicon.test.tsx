import Favicon from "./Favicon"

describe('SEO Favicon Component', () => { 
    it('renders with data properly', () => { 
        const result = Favicon({
            directory: ""
        }); 

        expect(result).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    rel: 'apple-touch-icon',
                    href: `/apple-icon-60x60.png`,
                    sizes: '60x60' 
                })
            ])
        );

        expect(result).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    rel: 'icon',
                    href: `/android-icon-192x192.png`,
                    sizes: '192x192' 
                })
            ])
        );

        expect(result).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    rel: 'icon',
                    href: `/favicon-32x32.png`,
                    sizes: '32x32'
                })
            ])
        );

        expect(result).toMatchSnapshot();
    });

    it('renders properly with the directory', () => { 
        const result = Favicon({
            directory: "/helloworld"
        }); 

        expect(result).toEqual(
            expect.arrayContaining([
                expect.objectContaining({
                    rel: 'apple-touch-icon',
                    href: `/helloworld/apple-icon-60x60.png`,
                    sizes: '60x60' 
                })
            ])
        );

        expect(result).toMatchSnapshot();
    });

    it('properly with options disabled', () => { 
        const result = Favicon({
            appleTouchIcon: false
        }); 

        expect(result).toEqual(
            expect.not.arrayContaining([
                expect.objectContaining({
                    rel: 'apple-touch-icon',
                    href: `/apple-icon-60x60.png`,
                    sizes: '60x60' 
                })
            ])
        );

        expect(result).toMatchSnapshot();
    });
});