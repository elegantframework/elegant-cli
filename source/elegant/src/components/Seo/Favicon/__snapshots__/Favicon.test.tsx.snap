// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO Favicon Component properly with options disabled 1`] = `
[
  {
    "href": "/favicons/android-icon-192x192.png",
    "rel": "icon",
    "sizes": "192x192",
  },
  {
    "href": "/favicons/manifest.json",
    "rel": "manifest",
  },
  {
    "href": "/favicons/favicon-16x16.png",
    "rel": "icon",
    "sizes": "16x16",
  },
  {
    "href": "/favicons/favicon-32x32.png",
    "rel": "icon",
    "sizes": "32x32",
  },
  {
    "href": "/favicons/favicon-96x96.png",
    "rel": "icon",
    "sizes": "96x96",
  },
]
`;

exports[`SEO Favicon Component renders properly with the directory 1`] = `
[
  {
    "href": "/helloworld/apple-icon-57x57.png",
    "rel": "apple-touch-icon",
    "sizes": "57x57",
  },
  {
    "href": "/helloworld/apple-icon-60x60.png",
    "rel": "apple-touch-icon",
    "sizes": "60x60",
  },
  {
    "href": "/helloworld/apple-icon-72x72.png",
    "rel": "apple-touch-icon",
    "sizes": "72x72",
  },
  {
    "href": "/helloworld/apple-icon-76x76.png",
    "rel": "apple-touch-icon",
    "sizes": "76x76",
  },
  {
    "href": "/helloworld/apple-icon-114x114.png",
    "rel": "apple-touch-icon",
    "sizes": "114x114",
  },
  {
    "href": "/helloworld/apple-icon-120x120.png",
    "rel": "apple-touch-icon",
    "sizes": "120x120",
  },
  {
    "href": "/helloworld/apple-icon-144x144.png",
    "rel": "apple-touch-icon",
    "sizes": "144x144",
  },
  {
    "href": "/helloworld/apple-icon-152x152.png",
    "rel": "apple-touch-icon",
    "sizes": "152x152",
  },
  {
    "href": "/helloworld/apple-icon-180x180.png",
    "rel": "apple-touch-icon",
    "sizes": "180x180",
  },
  {
    "href": "/helloworld/android-icon-192x192.png",
    "rel": "icon",
    "sizes": "192x192",
  },
  {
    "href": "/helloworld/manifest.json",
    "rel": "manifest",
  },
  {
    "href": "/helloworld/favicon-16x16.png",
    "rel": "icon",
    "sizes": "16x16",
  },
  {
    "href": "/helloworld/favicon-32x32.png",
    "rel": "icon",
    "sizes": "32x32",
  },
  {
    "href": "/helloworld/favicon-96x96.png",
    "rel": "icon",
    "sizes": "96x96",
  },
]
`;

exports[`SEO Favicon Component renders with data properly 1`] = `
[
  {
    "href": "/apple-icon-57x57.png",
    "rel": "apple-touch-icon",
    "sizes": "57x57",
  },
  {
    "href": "/apple-icon-60x60.png",
    "rel": "apple-touch-icon",
    "sizes": "60x60",
  },
  {
    "href": "/apple-icon-72x72.png",
    "rel": "apple-touch-icon",
    "sizes": "72x72",
  },
  {
    "href": "/apple-icon-76x76.png",
    "rel": "apple-touch-icon",
    "sizes": "76x76",
  },
  {
    "href": "/apple-icon-114x114.png",
    "rel": "apple-touch-icon",
    "sizes": "114x114",
  },
  {
    "href": "/apple-icon-120x120.png",
    "rel": "apple-touch-icon",
    "sizes": "120x120",
  },
  {
    "href": "/apple-icon-144x144.png",
    "rel": "apple-touch-icon",
    "sizes": "144x144",
  },
  {
    "href": "/apple-icon-152x152.png",
    "rel": "apple-touch-icon",
    "sizes": "152x152",
  },
  {
    "href": "/apple-icon-180x180.png",
    "rel": "apple-touch-icon",
    "sizes": "180x180",
  },
  {
    "href": "/android-icon-192x192.png",
    "rel": "icon",
    "sizes": "192x192",
  },
  {
    "href": "/manifest.json",
    "rel": "manifest",
  },
  {
    "href": "/favicon-16x16.png",
    "rel": "icon",
    "sizes": "16x16",
  },
  {
    "href": "/favicon-32x32.png",
    "rel": "icon",
    "sizes": "32x32",
  },
  {
    "href": "/favicon-96x96.png",
    "rel": "icon",
    "sizes": "96x96",
  },
]
`;
